File: cpp-tests/matrix_mult_classic.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <random>

using Matrix = std::vector<std::vector<double>>;

Matrix multiply(const Matrix& A, const Matrix& B, int n) {
    Matrix C(n, std::vector<double>(n, 0.0));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            for (int k = 0; k < n; ++k) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}

// Helper to generate a matrix with random values
Matrix generateMatrix(int n) {
    Matrix M(n, std::vector<double>(n));
    std::mt19937 gen(42);
    std::uniform_real_distribution<> dis(0.0, 1.0);
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            M[i][j] = dis(gen);
        }
    }
    return M;
}

int main() {
    std::vector<int> sizes = {100, 200, 300, 400, 500, 600, 700, 800, 900, 1000};
    
    std::cout << "Size,Time_ms" << "\n";
    
    for (int n : sizes) {
        Matrix A = generateMatrix(n);
        Matrix B = generateMatrix(n);
        
        auto start = std::chrono::high_resolution_clock::now();
        Matrix C = multiply(A, B, n);
        auto end = std::chrono::high_resolution_clock::now();
        
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
        std::cout << n << "," << duration.count() << "\n";
    }
    
    return 0;
}

File: go-tests/matrix_mult_classic.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func multiply(A, B [][]float64, n int) [][]float64 {
	C := make([][]float64, n)
	for i := range C {
		C[i] = make([]float64, n)
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				C[i][j] += A[i][k] * B[k][j]
			}
		}
	}
	return C
}

// Helper to generate a matrix with random values
func generateMatrix(n int) [][]float64 {
	M := make([][]float64, n)
	rand.Seed(42)
	for i := range M {
		M[i] = make([]float64, n)
		for j := range M[i] {
			M[i][j] = rand.Float64()
		}
	}
	return M
}

func main() {
	sizes := []int{100, 200, 300, 400, 500, 600, 700, 800, 900, 1000}

	fmt.Println("Size,Time_ms")

	for _, n := range sizes {
		A := generateMatrix(n)
		B := generateMatrix(n)

		start := time.Now()
		_ = multiply(A, B, n)
		duration := time.Since(start)

		fmt.Printf("%d,%d\n", n, duration.Milliseconds())
	}
}
